text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+trait which returns one instead of zero. We'll define it like this:
+
+```rust
+pub trait MyDefault: Sized {
+    fn my_default() -> Self;
+}
+```
+
+Right now, with what we know, we have to copy-paste our impl for every integer
+type. It would look something like:
+
+```rust
+impl MyDefault for u8 {
+    fn my_default() -> u8 { 1 }
+}
+```
+
+Except, we'd have to replicate this for every single integer type. If we're
+clever, we could use a macro:
+
+```rust
+macro_rules! impl_my_default {
+    ($($int: ident),*) => {$(
+        impl MyDefault for $int {
+            fn my_default() -> $int { 1 }
+        }
+    )*}
+}
+impl_my_default!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
+```
+
+But really, that's just making the compiler do the copy-pasting for us. Instead,
+we're going to use the special types `uint<N>` and `int<N>` to generalise the
+code for us. For this, we're going to use const generics, instead of type
+generics. The end result looks like:
+
+```rust
+impl<const N: usize> MyDefault for uint<N> {
+    fn my_default() -> uint<N> { 1 }
+}
+impl<const N: usize> MyDefault for int<N> {
+    fn my_default() -> int<N> { 1 }
+}
+impl MyDefault for usize {
+    fn my_default() -> usize { 1 }
+}
+impl MyDefault for isize {
+    fn my_default() -> isize { 1 }
+}
+```
+
+That's a lot better! As you'll notice, we still have to separately implement
+code for `usize` and `isize`; that's because they're separate types altogether.
+For example, on 64-bit systems, `usize` is technically just a `u64`, but
+`uint<64>` refers to `u64`, not `usize`. It's unfortunate, but we've reduced our
+number of impls from twelve to four, which is still a very big deal!
+
+One other side effect of having `uint<N>` and `int<N>` is that we can represent
+a lot more types than before. For example, `uint<7>` is just a seven-bit
+integer, which we might use to represent an ASCII character. Don't be fooled by
+these; we can't actually make `N` bigger than 128 (for now), and `uint<N>` is
+almost always stored using `uint<N.next_power_of_two()>`. `uint<7>` is stored as
+`u8`, `uint<10>` is stored as `u16`, `uint<48>` is stored as `u64`, etc. If you
+store a value with a nonstandard amount of bits, you're not actually saving any
+space, so, keep that in mind!
+
+That said, there are a few cases where we might want to use one of these
+weirdly sized types, as many data formats store values in terms of bits instead
+of bytes to save space. Eventually, we'll have a way of representing these, but
+not today.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Primitive behaviour
+
+The compiler will have two new built-in integer types: `uint<N>` and `int<N>`,
+where `const N: usize`. These will alias to existing `uN` and `iN` types if `N`
+is a power of two and no greater than 128. `usize` and `isize` remain separate
+types due to coherence issues, and `bool` remains separate from `uint<1>` as it
```

Comment:
Elaborate on what those coherence issues are for unfamiliar readers?

Reply:
Traits can be implemented separately & differently for u32, u64, usize, etc. so unifying usize with uN for appropriate N would cause overlapping impls.

Reply:
I may not be able to get to this by the weekend but I will try to remember to elaborate more on this.